src/main/java/io/airlift/compress/lz4/Lz4Native.java

```diff
@@ -0,0 +1,140 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.compress.lz4;
+
+import io.airlift.compress.internal.NativeLoader.Symbols;
+import io.airlift.compress.internal.NativeSignature;
+
+import java.lang.foreign.MemorySegment;
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static io.airlift.compress.internal.NativeLoader.loadSymbols;
+import static java.lang.invoke.MethodHandles.lookup;
+
+final class Lz4Native
+{
+    private Lz4Native() {}
+
+    private record MethodHandles(
+            @NativeSignature(name = "LZ4_compressBound", returnType = int.class, argumentTypes = int.class)
+            MethodHandle maxCompressedLength,
+            @NativeSignature(name = "LZ4_compress_fast", returnType = int.class, argumentTypes = {MemorySegment.class, MemorySegment.class, int.class, int.class, int.class})
+            MethodHandle compress,
+            @NativeSignature(name = "LZ4_compress_fast_extState", returnType = int.class, argumentTypes = {MemorySegment.class, MemorySegment.class, MemorySegment.class, int.class, int.class, int.class})
+            MethodHandle compressExternalState,
+            @NativeSignature(name = "LZ4_decompress_safe", returnType = int.class, argumentTypes = {MemorySegment.class, MemorySegment.class, int.class, int.class})
+            MethodHandle decompress,
+            @NativeSignature(name = "LZ4_sizeofState", returnType = int.class, argumentTypes = {})
+            MethodHandle sizeofState) {}
+
+    private static final Optional<LinkageError> LINKAGE_ERROR;
+    private static final MethodHandle MAX_COMPRESSED_LENGTH_METHOD;
+    private static final MethodHandle COMPRESS_METHOD;
+    private static final MethodHandle COMPRESS_EXTERNAL_STATE_METHOD;
+    private static final MethodHandle DECOMPRESS_METHOD;
+
+    // Defined in lz4.h: https://github.com/lz4/lz4/blob/v1.9.4/lib/lz4.c#L51
+    public static final int DEFAULT_ACCELERATION = 1;
+    public static final int STATE_SIZE;
+
+    static {
+        Symbols<MethodHandles> symbols = loadSymbols("lz4", MethodHandles.class, lookup());
+        LINKAGE_ERROR = symbols.linkageError();
+        MethodHandles methodHandles = symbols.symbols();
+        MAX_COMPRESSED_LENGTH_METHOD = methodHandles.maxCompressedLength();
+        COMPRESS_METHOD = methodHandles.compress();
+        COMPRESS_EXTERNAL_STATE_METHOD = methodHandles.compressExternalState();
+        DECOMPRESS_METHOD = methodHandles.decompress();
+        try {
+            STATE_SIZE = (int) methodHandles.sizeofState().invokeExact();
```

Comment:
this should be guarded with `if (LINKAGE_ERROR.isEmpty())`

so that we can call `Lz4Native.isEnabled()` to detect whether library loads properly. Right now if it doesn't it always fails and we cannot recover and fallback